### ISSUE TITLE
Metrics

### ISSUE THREAD

**Issue by hacdias** [ipfs-shipyard/ipfs-webui#1220]
@olizilla suggested we could add a metrics tab somewhere in the Web UI where we could show the data from `${apiAddr}/debug/metrics/prometheus` graphically. That would certainly be an interesting idea and useful for some kinds of users.

**Leaving this as a WIP Issue**

**Comment by SgtPooki** [comment]
> every go-ipfs/ipfs-desktop exposes them at http://localhost:5001/debug/metrics/prometheus
some of them appear only after initial activity happens (gateway request, tcp/quic traffic, storage operation etc) so initial list may be bit empty - @lidel

**Comment by SgtPooki** [comment]
@olizilla @hacdias @lidel I'm probably going to be taking this on, so I would love to get a few ideas from those who are interested in such a view. 

1. What questions are we trying to answer with this view? 
1. What problems do we expect this view to solve, or assist with solving?

**Comment by lidel** [comment]
My quick take is:

- we want to provide ipfs-webui (go-ipfs, ipfs-desktop) users  with a quick view for exploring metrics without the need for setting up [Expression Browser or Grafana](https://www.metricfire.com/blog/prometheus-dashboards/#Prometheus-Dashboard-Options).
- rationale: text is fine for basic `count`, but having useful visualization of `histogram` metrics makes it way easier to answer questions like:
  - is my gateway slow to respond, or the problem reported by the user is just an outlier? if so, how fast is 95th percent of responses?
  - what is the most popular gateway response type? which request type takes the most of time? 
  - how long most of quic/tcp connections last?
  - is there a visible degradation in the duration of my datastores' reads/writes?
  - etc

So in practice:

- Figure out how/where to add "Diagnostic" screen to ipfs-webui
  - Make it, so it can have multiple tabs (we want metrics, but we will also add logs, and other diagnostic tooling there over time, so good to create foundation for that from the start)
- "Metrics" tab should be a generic solution
  - Show it only when `debug/metrics/prometheus` endpoint is available, and provide UI for exploring metrics listed there.
  - Dynamic discovery: assume new metrics can be added over time, and old metrics can be removed or deprecated, so we can't expect anything to stay the same.
  -  Good news is that this wall of text is becoming a standard(?): 
     1. https://prometheus.io/docs/instrumenting/exposition_formats/#openmetrics-text-format
      2. https://github.com/OpenObservability/OpenMetrics
     3.  Given that each metric has standard `TYPE` like `counter` and `histogram` and unique name and description defined in `HELP` lines, we should be able to generate UI for each (if there is no JS tooling for this format already -- needs research)

**Comment by olizilla** [comment]
On my way to work this morning I started day-dreaming about how nice it would be to get back to working on developer tools for IPFS & co. A dangerous move as I was riding my bike at the time, but lo! What serendipitous timing! I'd love to pitch in on this!

That wall of text is the only format for getting that particular list of data points but it's intended for consumption by Prometheus, the timeseries db, but it's well standardised and much tooling exists for it. The numbers are all point-in-time measurements, intended to be scraped periodically, so it's not such a bad fit for a local dev tool that is already in the habit of polling the api for info.

There are some useful metrics in there that we use all the time now, to check if a node is working well, things like bitswap queue length, but in general much of the metrics are super specific, and only really of use to a developer who is making specific changes to a specific subsystem, so we should build up a short list of metrics that are worth putting the spotlight on, and then give experts a way to get to _the kitchen sink_.

**Comment by olizilla** [comment]
background concern:
- the data points available at debug/metrics/prometheus have grown organically over time... I'm not sure they are all useful or form a coherent overview of what is going on inside a go-ipfs node... my assumption is that the list represents a set of things that needed some performance tuning at some point.

>    What questions are we trying to answer with this view?
>    What problems do we expect this view to solve, or assist with solving?

This is a very good place to start... I don't have the answers, but it will involve talking to users and seeing what problems they are hitting.

In general, there is a need for high quality dev tooling for ipfs / libp2p / ipld ...when i proposed parsing info from `debug/metrics/prometheus`, I wasn't familiar with prometheus and was groping around in the dark for anything that might be interesting. 

Now I think we can use it to prototype something and get a feel for it, but we should be careful to not end up implementing an incidental subset of prometheus in js. Ideally we also use this work to improve the core ipfs diagnostic apis as well. I think as @SgtPooki is pointing to, parsing info from prometheus metrics for the web UI is not ideal.

**Comment by lidel** [comment]
Highly relevant work is happening in: 
- https://github.com/libp2p/go-libp2p/issues/1356
  - https://github.com/libp2p/go-libp2p-resource-manager/issues/50 

When we have it released and bubbled up to kubo (go-ipfs), ipfs-webui will have access to networking metrics around  resource usage (per libp2p protocol, service, peer).

**Comment by RamilMus** [comment]
Hello! 
Looked through your discussion and thought it is very relevant. Please tell me if I need to open a separate issue.
Can you please help with quic_connection_errors_total metric. When Prometheus scrapes the path   http://localhost:5001/debug/metrics/prometheus it gets thousands of lines with this error metric which are not combined, as a result prometheus marks this exporter as down due to time out. Is there a way to disable them?
